Python/pythonrun.c

```diff
@@ -280,11 +280,44 @@ PyRun_InteractiveOneObjectEx(FILE *fp, PyObject *filename,
     PyObject *main_dict = PyModule_GetDict(main_module);  // borrowed ref
 
     PyObject *res = run_mod(mod, filename, main_dict, main_dict, flags, arena, interactive_src, 1);
+    Py_INCREF(interactive_src);
     _PyArena_Free(arena);
     Py_DECREF(main_module);
     if (res == NULL) {
+        PyThreadState *tstate = _PyThreadState_GET();
+        PyObject *exc = _PyErr_GetRaisedException(tstate);
+        if (PyType_IsSubtype(Py_TYPE(exc),
+                             (PyTypeObject *) PyExc_SyntaxError))
+        {
+            /* fix "text" attribute */
+            assert(interactive_src != NULL);
+            PyObject *xs = PyUnicode_Splitlines(interactive_src, 1);
+            if (xs == NULL) {
+                goto err;
+            }
+            PyObject *ln = PyObject_GetAttr(exc, &_Py_ID(lineno));
+            if (ln == NULL) {
+                Py_DECREF(xs);
+                goto err;
+            }
+            int n = PyLong_AsInt(ln);
+            Py_DECREF(ln);
+            if (n <= 0 || n > PyList_GET_SIZE(xs)) {
```

Comment:
If you have an overflow, don't you need to somehow clear the exception's state (maybe because there is some reference to the exception?) I *think* it would otherwise replace the current exception (or am I wrong?)

Reply:
Good catch.

Reply:
But in this case we just reset raised exception on L317.

Reply:
My idea was

```C
if (n < 0) {
	// clears the overflow error so that we correctly decref
	Py_DECREF(xs);
	_PyErr_Clear(tstate);
	goto error;
}
else if (n > PyList_GET_SIZE(xs)) {
	Py_DECREF(xs);
	// nothing to clear since _PyErr_GetRaisedException() did it already
	goto error;
}
```

This could be simplified into 

```C
if (n < 0 || n > PyList_GET_SIZE(xs)) {
	Py_DECREF(xs);
	_PyErr_Clear(tstate);
	goto error;
}
```

since `_PyErr_Clear()` is a no-op if the error indicator is not set.

Reply:
(This is what you were actually doing in the check for `PyObject_SetAttr()`)

Reply:
> // clears the overflow error so that we correctly decref

Actually, condition n < 0 could be without any exception set.

Correct decref happens in _PyErr_SetRaisedException(tstate, exc), if new exception was set.